assignment-client/src/assets/AssetServer.cpp

```diff
@@ -0,0 +1,279 @@
+    } else {
+        qDebug() << "Asset not found: " << assetHash;
+        replyPacket->writePrimitive(AssetServerError::ASSET_NOT_FOUND);
+    }
+
+    auto nodeList = DependencyManager::get<NodeList>();
+    nodeList->sendPacket(std::move(replyPacket), *senderNode);
+}
+
+void AssetServer::handleAssetGet(QSharedPointer<NLPacket> packet, SharedNodePointer senderNode) {
+    if (packet->getPayloadSize() < HASH_HEX_LENGTH) {
+        qDebug() << "ERROR bad file request";
+        return;
+    }
+
+    MessageID messageID;
+    packet->readPrimitive(&messageID);
+
+    QByteArray assetHash;
+    assetHash = packet->read(HASH_HEX_LENGTH);
+
+    DataOffset start;
+    packet->readPrimitive(&start);
+
+    DataOffset end;
+    packet->readPrimitive(&end);
+
+    qDebug() << "Received a request for the file: " << assetHash << " from " << start << " to " << end;
+
+    // Queue task
+    QString filePath = _resourcesDirectory.filePath(QString(assetHash));
+    auto task = new SendAssetTask(messageID, assetHash, filePath, start, end, senderNode);
+    _taskPool.start(task);
+}
+
+void AssetServer::handleAssetUpload(QSharedPointer<NLPacketList> packetList, SharedNodePointer senderNode) {
+    auto data = packetList->getMessage();
+    QBuffer buffer { &data };
+    buffer.open(QIODevice::ReadOnly);
+
+    MessageID messageID;
+    buffer.read(reinterpret_cast<char*>(&messageID), sizeof(messageID));
+    // buffer.readPrimitive(&messageID);
+
+    uint8_t extensionLength;
+    buffer.read(reinterpret_cast<char*>(&extensionLength), sizeof(extensionLength));
+    // buffer.readPrimitive(&extensionLength);
+
+    QByteArray extension = buffer.read(extensionLength);
+
+    qDebug() << "Got extension: " << extension;
+
+    uint64_t fileSize;
+    buffer.read(reinterpret_cast<char*>(&fileSize), sizeof(fileSize));
+    // buffer.readPrimitive(&fileSize);
+
+    // const uint64_t MAX_LENGTH = 1024;
+    // fileSize = std::min(MAX_LENGTH, fileSize);
+    qDebug() << "Receiving a file of size " << fileSize;
+
+    QByteArray fileData = buffer.read(fileSize);
+
+    QString hash = hashData(fileData);
+
+    qDebug() << "Got data: (" << hash << ") ";
+
+    QFile file { _resourcesDirectory.filePath(QString(hash)) }; //+ "." + QString(extension) };
+
+    if (file.exists()) {
+        qDebug() << "[WARNING] This file already exists";
+    } else {
+        file.open(QIODevice::WriteOnly);
+        file.write(fileData);
+        file.close();
+    }
+
+    auto replyPacket = NLPacket::create(PacketType::AssetUploadReply);
+
+    replyPacket->writePrimitive(messageID);
+
+    replyPacket->writePrimitive(true);
+    replyPacket->write(hash.toLatin1().constData(), HASH_HEX_LENGTH);
+
+    auto nodeList = DependencyManager::get<NodeList>();
+    nodeList->sendPacket(std::move(replyPacket), *senderNode);
+}
+
+QString AssetServer::hashData(const QByteArray& data) {
+    return QString(QCryptographicHash::hash(data, QCryptographicHash::Md5).toHex());
```

Comment:
Think we probably want this to be QCryptographicHash::Sha256 - there are collisions in Md5